time_weighted_average_price/attest_fn_call/README.md

```diff
@@ -16,37 +23,473 @@ function call to expand the set of samples.
+	NumSamples   int                       `json:"num_samples"`
+	EAttest      json.RawMessage           `json:"eAttest"`
+	TAttest      json.RawMessage           `json:"tAttest"`
+	Whitelist    []basm.EnclaveMeasurement `json:"whitelist"`
+}
+
+//export iteration
+func iteration(inputPtr, secretPtr uint64) uint64 {
+	var args ArgsIterate
+	inputData := basm.ReadFromHost(inputPtr)
+	err := json.Unmarshal(inputData, &args)
+	if err != nil {
+		outErr := fmt.Errorf("could not unmarshal args args: %w", err)
+		return WriteError(outErr)
+	}
+
+	priceSamples, err := extractPriceSamples(args.EAttest, args.TAttest, args.Whitelist)
+	if err != nil {
+		outErr := fmt.Errorf("extracting priceSamples: %w", err)
+		return WriteError(outErr)
+	}
+
+	newPriceSample, err := getNewPriceSample(args.TokenAddress, args.ChainID)
+	if err != nil {
+		outErr := fmt.Errorf("getting new sample: %w", err)
+		return WriteError(outErr)
+	}
+
+	nextPriceSamples := append(priceSamples, newPriceSample)
+	if len(nextPriceSamples) > args.NumSamples {
+		numToRemove := len(nextPriceSamples) - args.NumSamples
+		nextPriceSamples = nextPriceSamples[numToRemove:]
+	}
+
+	return WriteOutput(nextPriceSamples)
+}
+```
+
+where we: 
+
+1. Extract `iteration` call arguments `args` from the `inputPtr`.
+2. Call `extractPriceSamples` to extract the previous collection of price
+   samples `prevPriceSamples` from the transitively attested function call
+   `args.TAttest`.
+3. Collects a new price sample by calling `getNewPriceSample`.
+4. Compute the new collection of price samples `nextPriceSamples` by appending
+   the new price sample to the previous collection and then truncating the
+   collection to `args.NumSamples` most recent samples.
+5. Return `nextPriceSamples` to the Blocky AS server for attestation.
+
+Let's dive a bit deeper into is the `extractPriceSamples` function:
+
+```go
+func extractPriceSamples(
+	eAttest json.RawMessage,
+	tAttest json.RawMessage,
+	whitelist []basm.EnclaveMeasurement,
+) (
+	[]price.Price,
+	error,
+) {
+	// bootstrap with empty samples if we don't have a transitive attestation
+	if tAttest == nil {
+		return []price.Price{}, nil
+	}
+
+	verifiedTA, err := basm.VerifyAttestation(
+		basm.VerifyAttestationInput{
+			EnclaveAttestedKey:       eAttest,
+			TransitiveAttestedClaims: tAttest,
+			AcceptableMeasures:       whitelist,
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not verify previous attestation: %w", err)
+	}
+
+	verifiedClaims, err := xbasm.ParseFnCallClaims(verifiedTA.RawClaims)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse claims: %w", err)
+	}
+
+	var prevResult Result
+	err = json.Unmarshal(verifiedClaims.Output, &prevResult)
+	switch {
+	case err != nil:
+		return nil, fmt.Errorf("could not unmarshal previous output: %w", err)
+	case !prevResult.Success:
+		return nil, fmt.Errorf("previous run was an error: %w", err)
+	}
+
+	prevPriceSamplesStr, err := json.Marshal(prevResult.Value)
```

Comment:
Not saying that you need to do this, however, this marshal to unmarshal thing is kind of confusing (and is kind of a hack).  I do understand that it is non-trivial to go from `any` to `[]Price` we could do it explicitly, however, I am wondering if this provides a pretty good argument for using a result type that uses generics. 
I put together an example on a go playground https://go.dev/play/p/PzS13xL-oGA